Network/Socket/Shutdown.hs

```diff
@@ -34,3 +40,40 @@ shutdown s stype = void $ withFdSocket s $ \fd ->
 
 foreign import CALLCONV unsafe "shutdown"
   c_shutdown :: CInt -> CInt -> IO CInt
+
+
+-- | Closing a socket gracefully.
+--   This sends TCP FIN and check if TCP FIN is received from the peer.
+--   The second argument is time out to receive TCP FIN in millisecond.
+--   In both normal cases and error cases, socket is deallocated finally.
+--
+--   Since: 3.1.1.0
+gracefulClose :: Socket -> Int -> IO ()
+gracefulClose s tm = (sendRecvFIN `E.finally` close s) `E.catch` ignore
+  where
+    sendRecvFIN = do
+        -- Sending TCP FIN.
+        shutdown s ShutdownSend
+        -- Waiting TCP FIN.
+        E.bracket (mallocBytes bufSize) free recvEOF
+    recvEOF buf = loop 0
+      where
+        loop delay = do
+            -- We don't check the (positive) length.
+            -- In normal case, it's 0. That is, only FIN is received.
+            -- In error cases, data is available. But there is no
+            -- application which can read it. So, let's stop receiving
+            -- to prevent attacks.
+            r <- recvBufNoWait s buf bufSize
+            let delay' = delay + clock
+            when (r == -1 && delay' < tm) $ do
+                threadDelay (clock * 1000)
+                loop delay'
+    -- Don't use 4092 here. The GHC runtime takes the global lock
+    -- if the length is over 3276 bytes in 32bit or 3272 bytes in 64bit.
+    bufSize = 1024
+    -- milliseconds. Taken from BSD fast clock value.
+    clock = 200
+    -- shutdown sometime returns ENOTCONN.
+    -- Probably, we don't want to log this error.
+    ignore (E.SomeException _e) = return ()
```

Comment:
I'd advise using `E.IOException` instead to avoid ignoring async exceptions. Especially given the usage of `threadDelay`, it's reasonable that an external thread may end up trying to kill this operation.

Reply:
* Here you're ignoring _all_ exceptions, not only `ENOTCONN`; we'd have to check `errno` for that.
* While it _may_ be acceptable that `close` ignores exceptions (see https://github.com/haskell/network/blob/bbb66aaf4d88d53cbff652b3588bf9fa5745e3ce/Network/Socket/Types.hsc#L169-L170), I don't think it's OK here, because you're `catch`ing this not only over `close s` but also `sendRecvFIN`. I don't think we should ingore exceptions of `sendRecvFIN`. It calls `shutdown`, which has proper `errno` raising behaviour, and we should re-raise those exceptions. At least all the ones that aren't ENOTCONN`, but even in that case I'm not sure why it's legitimate to not throw on `ENOTCONN`; the comment probably needs expansion. The user could also call accidentally call `gracefulClose` on an unconnected socket; wouldn't we want to tell them that they did something wrong by raising the `ENOTCONN`?
* Somewhat unrelatedly, @snoyberg and I think that even `close` should probably not silently discard `EBADF` as it currently does, because that's a user error. (The other `errno`s leave the socket in undefined state so there it's probably more OK.)

Reply:
@snoyberg You are right. As a quick response, I changed it to `IOExeption`. But I should consider @nh2's comments later.

Reply:
@nh2 I will consider error handling for `gracefulClose` more. Error handling for other APIs should be discussed in #151.

Reply:
@nh2 It seems to me that my experience was based on the previous *wrong* implementation using `unsafe` calls. Using the current implementation, I don't need to ignore `ENOTCONN` anymore since it does not occur in tests. So, let's remove `ignore`.